README.md

```diff
@@ -291,173 +321,204 @@ be used in global variable initializers.
+   codepoints will have *N*+1 distinct valid WTF-8 position values: one
+   before each codepoint, and one at the end.  A position value of 0
+   denotes the string start (before the first codepoint).  It follows
+   that 0 may also denote the end of the string also, for zero-length
+   strings.
+
+ * WTF-16 positions address 16-bit code units.  A string consisting of
+   *N* code units when encoded as WTF-16 will have *N*+1 distinct valid
+   position: one before each 16-bit code unit, and one at the end.  As
+   with WTF-8 positions, 0 denotes the string start.  There may be more
+   WTF-16 positions in a string than there are WTF-8 positions, as
+   there may be multiple WTF-16 code units per codepoint.
+   A WTF-16 position is the same as the operand to JavaScript's
+   *[`String.prototype.charCodeAt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt).
+
+A position is an offset into a string's contents.  Positions are
+strictly ordered, and therefore can be compared against each other.
+
+We expect WebAssembly implementations to represent strings using either
+WTF-8 or WTF-16, and thus one of these encodings is "native" and the
+other is "foreign".  Some implementations will want to use
+[breadcrumbs](https://www.swift.org/blog/utf8-string/#breadcrumbs) to
+project foreign positions to native positions.  A simple one-entry cache
+may also suffice for some implementations.  Finally, we expect that many
+source languages will process strings in chunks via in-memory encoding,
+minimizing per-codepoint translation cost between foreign and native
+offsets.
+
+For WTF-16 positions, since position values are packed, we just provide
+an accessor to get the last position in the string:
+
+```
+(string.end_wtf16 str:stringref)
+  -> pos:i32
+```
+Return the position after the last code unit in the WTF-16 encoding of
+the string *str*.
 
-#### Invalid cursor values
+For WTF-8 positions, use the `string.advance_wtf8` and
+`string.rewind_wtf8` seek instructions:
 
-All instructions which take cursors trap if the cursor is not valid for
-the given string.
+```
+(string.advance_wtf8 str:stringref pos:i32 codepoints:i32)
+  -> pos:i32, codepoints:i32
+(string.rewind_wtf8 str:stringref pos:i32 codepoints:i32)
+  -> pos:i32, codepoints:i32
+```
+Return a new WTF-8 position for the string *`str`* which is
+*`codepoints`* codepoints after or before *`pos`*.  Also return the
+number of codepoints by which the new position differs from the old,
+which may be less than the *`codepoints`* parameter if end-of-string or
+beginning-of-string would be reached.  The *`codepoints`* argument is
+interpreted as a `u32`.
 
-#### Host strings which are not valid USV sequences
+#### Invalid position values
 
-In this MVP, it is not possible to produce a string which is not a valid
-sequence of unicode scalar values.
+All instructions which take string positions trap if the position is not
```

Comment:
I'm worried that this might be expensive.
IIUC, it's not linear in complexity, because in both WTF-16 and WTF-8 code units can be classified in O(1) as being a non-first element of a surrogate pair/chain or not. Still, there's a cost to doing this check for every access. We might just have to prototype it and see if that cost is acceptable.

Reply:
Good question!  Related to #10.

I think that for WTF-16 things are easy -- a valid position is one that's in range.  I think we have to check for this already so there's no additional work.  Do lmk if I am not thinking straight.  Note that the WTF-16 interfaces index over code units and not codepoints -- i.e. there's no special requirement forbidding a position value which is the offset of the low surrogate of a surrogate pair.

For WTF-8, we have the range check (no additional work even on WTF-16 hosts) and then the codepoint check.  Agreed that this is a cost.  Let's keep thinking about this in #10.